File: src/2025-02-27-deepclaude37-rust/python/exercises/practice/tree-building/tree_building.py
class Record:
    def __init__(self, record_id, parent_id):
        self.record_id = record_id
        self.parent_id = parent_id


class Node:
    def __init__(self, node_id):
        self.node_id = node_id
        self.children = []


def BuildTree(records):
    """Build a tree from the given records."""
    if not records:
        return None
        
    # Sort and validate records
    records.sort(key=lambda x: x.record_id)
    
    # Check if record IDs are continuous from 0 to n-1
    ordered_ids = [record.record_id for record in records]
    if ordered_ids != list(range(len(records))):
        raise ValueError("Record id is invalid or out of order.")
    
    # Create nodes using a dictionary for O(1) lookups
    nodes = {}
    for record in records:
        # Create a new node for this record
        nodes[record.record_id] = Node(record.record_id)
        
        # Check root node
        if record.record_id == 0:
            if record.parent_id != 0:
                raise ValueError("Node parent_id should be smaller than it's record_id.")
            continue
            
        # Validate parent relationship
        if record.parent_id >= record.record_id:
            if record.parent_id == record.record_id:
                raise ValueError("Only root should have equal record and parent id.")
            else:
                raise ValueError("Node parent_id should be smaller than it's record_id.")
            
        # Add this node as a child to its parent
        nodes[record.parent_id].children.append(nodes[record.record_id])
    
    # Return the root node (if any nodes exist)
    return nodes[0] if nodes else None

File: src/2025-02-27-deepclaude37-rust/go/exercises/practice/tree-building/tree_building.go
package tree

import (
	"fmt"
	"sort"
)

type Record struct {
	ID     int
	Parent int
	// feel free to add fields as you see fit
}

type Node struct {
	ID       int
	Children []*Node
	// feel free to add fields as you see fit
}

func Build(records []Record) (*Node, error) {
	if len(records) == 0 {
		return nil, nil
	}

	// Sort records by ID for processing in order
	sort.Slice(records, func(i, j int) bool {
		return records[i].ID < records[j].ID
	})

	// Validate records
	for i, r := range records {
		// Check IDs are contiguous and start from 0
		if r.ID != i {
			return nil, fmt.Errorf("non-contiguous or invalid ID: %d", r.ID)
		}

		// Check parent relationships
		if i == 0 {
			// Root node must have parent = self
			if r.Parent != 0 {
				return nil, fmt.Errorf("root node has invalid parent: %d", r.Parent)
			}
		} else {
			// Non-root nodes must have parent ID < own ID
			if r.Parent >= r.ID {
				return nil, fmt.Errorf("node %d has invalid parent: %d", r.ID, r.Parent)
			}
		}
	}

	// Create nodes
	nodes := make([]*Node, len(records))
	for i := range records {
		nodes[i] = &Node{ID: i}
	}

	// Build tree structure
	for i := 1; i < len(records); i++ {
		parentID := records[i].Parent
		nodes[parentID].Children = append(nodes[parentID].Children, nodes[i])
	}

	return nodes[0], nil
}
